File: _etc/Java/4963 섬의 개수.java
import java.util.Scanner;
import java.util.*;




public class Main {
    static int w, h;
    static int[][] map = new int[51][51];
    static boolean[][] visited = new boolean[51][51];
    static int[] dx = {-1, -1, 0, 1, 1, 1, 0, -1};
    static int[] dy = {0, 1, 1, 1, 0, -1, -1, -1};
    
    static boolean BFS(int x, int y) {
        
        if(map[x][y] == 0 || visited[x][y] == true) return false;
        int[] start = {x, y};
        Queue<int[]> queue = new LinkedList<int[]>();
        
        queue.add(start);
        
        while(queue.size() != 0) {
            
            int[] v = queue.poll();
            
            for(int i = 0; i < 8; i++) {
                int nx = dx[i] + v[0];
                int ny = dy[i] + v[1];
                
                if(nx < 0 || ny < 0 || nx >= h || ny >= w) continue;
                
                if(map[nx][ny] == 1 && visited[nx][ny] == false) {
                    int[] index = {nx, ny};
                    queue.add(index);
                    visited[nx][ny] = true;
                }
            }

        }
        
        return true;
    }
    static void init() {
        for(int i = 0 ; i < map.length; i++) {
            Arrays.fill(map[i], 0);
        }
        for(int i = 0 ; i < visited.length; i++) {
            Arrays.fill(visited[i], false);
        }
    }
    
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        int answer = 0;
        while(true) {
            answer = 0;
            w = sc.nextInt(); h = sc.nextInt();
            if(w == 0 && h == 0) break;
            
            for(int i = 0 ; i < h; i++) {
                for(int j = 0; j < w; j++) {
                    int info = sc.nextInt();
                    map[i][j] = info;
                }
            }
            
            
            for(int i = 0 ; i < h; i++) {
                for(int j = 0; j < w; j++) {
                    if(BFS(i,j)) answer += 1;
                }
            }
            System.out.println(answer);
            init();
        }
    }
}

File: _etc/C#/4963 섬의 개수.cs
using System;
using static System.Console;
using System.Collections.Generic;

class Program
{
    static int w, h;
    static int[,] board = new int[51,51];
    static bool[,] visited = new bool[51,51];
    static int[] dx = {-1, -1, 0 ,1, 1 ,1, 0, -1};
    static int[] dy = {0, 1, 1, 1, 0, -1, -1, -1};
    
    
    static bool BFS(Tuple<int, int> start) {
        
        if(board[start.Item1, start.Item2] == 0 || visited[start.Item1, start.Item2] == true) return false;
        
        Queue<Tuple<int, int>> queue = new Queue<Tuple<int, int>>();
        queue.Enqueue(start);
        visited[start.Item1,start.Item2] = true;
        
        
        while(queue.Count != 0 ) {
            var v = queue.Dequeue();
            for(int i = 0 ; i < 8; i++) {
                int nx = v.Item1 + dx[i];
                int ny = v.Item2 + dy[i];
                
                
                if(nx < 0 || ny < 0 || nx >= h || ny >= w) continue;
                if(board[nx,ny] == 1 && visited[nx,ny] == false) {
                    visited[nx,ny] = true;
                    queue.Enqueue(new Tuple<int, int>(nx, ny));
                }
                
            }
            
        }
        return true;
    }
    
    static void InitBoard() {
        board = new int[51,51];
        visited = new bool[51,51];
    }
    
    static void Main() {
        while(true) {
            int answer = 0;
            string s = Console.ReadLine();
            string[] sa = s.Split();
            w = int.Parse(sa[0]); h = int.Parse(sa[1]);
            if(w == 0 && h == 0) break;
            
            for(int i = 0 ; i < h; i++) {
                string info = Console.ReadLine();
                string[] infoArr = info.Split(' ');
                for(int j = 0 ; j < w; j++) {
                    board[i,j] = int.Parse(infoArr[j]);
                }
            }
            
            
            for(int i = 0 ; i < h; i++) {
                for (int j = 0 ; j < w; j++) {
                    if(BFS(new Tuple<int,int>(i,j))) answer += 1;
                }
            }
            
            
            Console.WriteLine(answer);
            InitBoard();
        }
    }
}
